invoices/sql_migration.go

```diff
@@ -0,0 +1,557 @@
+			if !ok {
+				return fmt.Errorf("missing htlc for AMP htlc: "+
+					"%v", circuitKey)
+			}
+
+			params := sqlc.InsertAMPSubInvoiceHTLCParams{
+				InvoiceID:  invoiceID,
+				SetID:      setID[:],
+				HtlcID:     sqlHtlcID,
+				RootShare:  rootShare[:],
+				ChildIndex: int64(htlc.AMP.Record.ChildIndex()),
+				Hash:       htlc.AMP.Hash[:],
+			}
+
+			if htlc.AMP.Preimage != nil {
+				params.Preimage = htlc.AMP.Preimage[:]
+			}
+
+			err = tx.InsertAMPSubInvoiceHTLC(ctx, params)
+			if err != nil {
+				return fmt.Errorf("unable to insert AMP sub "+
+					"invoice: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// OverrideInvoiceTimeZone overrides the time zone of the invoice to the local
+// time zone and chops off the nanosecond part for comparison. This is needed
+// because KV database stores times as-is which as an unwanted side effect would
+// fail migration due to time comparison expecting both the original and
+// migrated invoices to be in the same local time zone and in microsecond
+// precision. Note that PostgreSQL stores times in microsecond precision while
+// SQLite can store times in nanosecond precision if using TEXT storage class.
+func OverrideInvoiceTimeZone(invoice *Invoice) {
+	fixTime := func(t time.Time) time.Time {
+		return t.In(time.Local).Truncate(time.Microsecond)
+	}
+
+	invoice.CreationDate = fixTime(invoice.CreationDate)
+
+	if !invoice.SettleDate.IsZero() {
+		invoice.SettleDate = fixTime(invoice.SettleDate)
+	}
+
+	if invoice.IsAMP() {
+		for setID, ampState := range invoice.AMPState {
+			if ampState.SettleDate.IsZero() {
+				continue
+			}
+
+			ampState.SettleDate = fixTime(ampState.SettleDate)
+			invoice.AMPState[setID] = ampState
+		}
+	}
+
+	for _, htlc := range invoice.Htlcs {
+		if !htlc.AcceptTime.IsZero() {
+			htlc.AcceptTime = fixTime(htlc.AcceptTime)
+		}
+
+		if !htlc.ResolveTime.IsZero() {
+			htlc.ResolveTime = fixTime(htlc.ResolveTime)
+		}
+	}
+}
+
+// MigrateInvoicesToSQL runs the migration of all invoices from the KV database
+// to the SQL database. The migration is done in a single transaction to ensure
+// that all invoices are migrated or none at all. This function can be run
+// multiple times without causing any issues as it will check if the migration
+// has already been performed.
+func MigrateInvoicesToSQL(ctx context.Context, db kvdb.Backend,
+	kvStore InvoiceDB, tx *sqlc.Queries, batchSize int) error {
+
+	log.Infof("Starting migration of invoices from KV to SQL")
+
+	offset := uint64(0)
+	t0 := time.Now()
+	// Create the hash index which we will use to look up invoice
+	// payment hashes by their add index during migration.
+	err := createInvoiceHashIndex(ctx, db, tx)
+	if err != nil && !errors.Is(err, ErrNoInvoicesCreated) {
+		log.Errorf("Unable to create invoice hash index: %v",
+			err)
+
+		return err
+	}
+	log.Debugf("Created SQL invoice hash index in %v", time.Since(t0))
+
+	total := 0
+	// Now we can start migrating the invoices. We'll do this in
```

Comment:
just one more question that came to mind post-review:

does this actually reduce a bunch of memory usage since on the SQL DB side, it will still take a bunch of memory due to this all being in a single tx.. right?

iirc, we've done previous migrations where we do 1 tx per batch and then persist a check mark in case things go wrong.  If anything goes wrong, we can just clear the tables and start again on the next attempt.

Reply:
the other thing is: if the migration succeeds successfully, should we not clear the "invoices" bucket in the kv store?